metrics-operator/controllers/analysis/controller.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package analysis
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/go-logr/logr"
+	metricsapi "github.com/keptn/lifecycle-toolkit/metrics-operator/api/v1alpha3"
+	common "github.com/keptn/lifecycle-toolkit/metrics-operator/controllers/common/analysis"
+	"golang.org/x/exp/maps"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// AnalysisReconciler reconciles an Analysis object
+type AnalysisReconciler struct {
+	client.Client
+	Scheme     *runtime.Scheme
+	Log        logr.Logger
+	MaxWorkers int //maybe 2 or 4 as def
+	Namespace  string
+	NewWorkersPoolFactory
+	common.IAnalysisEvaluator
+}
+
+//+kubebuilder:rbac:groups=metrics.keptn.sh,resources=analyses,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=metrics.keptn.sh,resources=analyses/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=metrics.keptn.sh,resources=analyses/finalizers,verbs=update
+// +kubebuilder:rbac:groups=metrics.keptn.sh,resources=keptnmetricsproviders,verbs=get;list;watch;
+//+kubebuilder:rbac:groups=metrics.keptn.sh,resources=analysisdefinitions,verbs=get;list;watch;
+//+kubebuilder:rbac:groups=metrics.keptn.sh,resources=analysisvaluetemplates,verbs=get;list;watch;
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// For more details, check Reconcile and its AnalysisResult here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (a *AnalysisReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	a.Log.Info("Reconciling Analysis")
+	analysis := &metricsapi.Analysis{}
+
+	//retrieve analysis
+	if err := a.Client.Get(ctx, req.NamespacedName, analysis); err != nil {
+		if errors.IsNotFound(err) {
+			// taking down all associated K8s resources is handled by K8s
+			a.Log.Info("Analysis resource not found. Ignoring since object must be deleted")
+			return ctrl.Result{}, nil
+		}
+		a.Log.Error(err, "Failed to get the Analysis")
+		return ctrl.Result{}, err
```

Comment:
```suggestion
		return ctrl.Result{}, nil
```

Reply:
why not return an error here?

Reply:
it's an error of fetching analysis, I guess it's enough to log the error, we have the same approach in metrics controller and probably also the rest of

Reply:
I believe this is something we will need to fix in the other controllers where we do not return an error in this case then. But I do not see any reason why we should not return the error here tbh

Reply:
ok, let's keep it as it is

Reply:
IMHO we MUST return an error. If we don't return an error, the reconciliation is considered successful and we won't try again. It could be that we get an error due to API rate limits of ETCD and, therefore, we should let the retry kick in.